### ISSUE TITLE
GPT model output nans log-likelihood when loading fp16 weights

### ISSUE THREAD

**Issue by Terrencezzj** [NVIDIA/FasterTransformer#221]
Hi,

We followed the pytorch example to run the gpt model.

The gpt models load weights in np.single and convert them to half. Now we want to load fp16 weights directly, but the output of the model became nothing. We checked the llhs and found they are nans. We also checked the loaded weights (both on the pytroch side and cpp ops file), they are successfully loaded, and the digits are the same as before. 

We are using v5.0_beta.

Any ideas on the possible causes?

**Comment by byshiue** [comment]
How do you load the fp16 weight directly? Note that we have some special reshape in the checkpoint converter.

**Comment by Terrencezzj** [comment]
> How do you load the fp16 weight directly? Note that we have some special reshape in the checkpoint converter.

https://github.com/cohere-ai/FasterTransformer/blob/6fddeac5f59ce4df380002aa945da57a0c8e878c/examples/pytorch/gpt/utils/gpt.py#L138

Like we changed dtype to np.float16 in this line.

**Comment by byshiue** [comment]
The page is 404 from my side.

**Comment by Terrencezzj** [comment]
> The page is 404 from my side.

sry, updated the code link.

https://github.com/NVIDIA/FasterTransformer/blob/7d17fda720b3494149444331440f1fe1544076de/examples/pytorch/gpt/utils/gpt.py#L138

**Comment by byshiue** [comment]
When you use 
```
np.fromfile(filename, dtype=np.float16)
```
your weights should be stored in float16. But in default, these weights are stored in float32.
Try to load like
```
np.fromfile(filename, dtype=np.float32).astype(np.float16)
```

**Comment by Terrencezzj** [comment]
> When you use
> 
> ```
> np.fromfile(filename, dtype=np.float16)
> ```
> 
> your weights should be stored in float16. But in default, these weights are stored in float32. Try to load like
> 
> ```
> np.fromfile(filename, dtype=np.float32).astype(np.float16)
> ```

thanks for the quick reply. we had saved our weights in float16. And loaded them like this, checked the weights after loading. The weights' numbers are correct, the same as before. We even printed out the passed weights in ParallelGptOp.h, its still the same digits as before. so we have no idea where could be wrong.

**Comment by byshiue** [comment]
What's the meaning for "but the output of the model became nothing"? And how you get the log likelihood? FT does not support to return it. 

You can also try the latest main branch.

**Comment by Terrencezzj** [comment]
Thanks for the suggestion. yes, we added the function of returning llh. We also printed out the logits in cuda and found they became nans. We have not located where they started to be nans. Will update if we make some progress.

**Comment by byshiue** [comment]
You can print the values of tensor of transformer layer to check where is broken. 
Some tools are in `cuda_utils.h`, like `print_abs_mean` and `print_to_screen`. They may help to locate the problem.

**Comment by Terrencezzj** [comment]
thanks! this would definitely be helpful.
